library/proc_macro/src/lib.rs

```diff
@@ -1503,7 +1503,8 @@ pub mod tracked_env {
     #[unstable(feature = "proc_macro_tracked_env", issue = "99515")]
     pub fn var<K: AsRef<OsStr> + AsRef<str>>(key: K) -> Result<String, VarError> {
         let key: &str = key.as_ref();
-        let value = env::var(key);
+        let value = crate::bridge::client::FreeFunctions::injected_env_var(key)
+            .map_or_else(|| env::var(key), Ok);
         crate::bridge::client::FreeFunctions::track_env_var(key, value.as_deref().ok());
```

Comment:
Is it correct to track `--env` here? I don't think it is. Tracked env vars end up in depinfo, implying that the compiler invocation depended on this environment variable, but it didn't, it just depended on a flag, which exists already. I think it should only track when `env::var` is used.

Reply:
Well, if this function is called, it means the environment variable was used no? Or did I misunderstand what you meant?

Reply:
No "environment variable" was used. It just took a value for the flag. A build system doesn't have to check whether an environment variables changes to figure out whether a rebuild is needed, it needs to look at arguments instead (which it of course already does).

Reply:
I just checked your previous PR and I missed that there:
https://github.com/rust-lang/rust/pull/118368/files#diff-0eaa1095d59f2f3d9c015005d15fd8f7cda13121a8404d77b2955fc5fcd40449R37

The name here also makes it more clear what tracking is for: it's for depinfo which is used by builds systems or other consumers to figure out what went into the compilation (for example when things need to be rebuilt). It's essential that rustc records env variables it reads with `env::var`. But `--env` are not actually env variables in the build environment, so they don't need to be tracked. I'd also be happy to review a PR changing the behavior for the macros to not track when the var is read from `--env`>

Reply:
I disagree with that. If you change `--env` for a used variable, it needs to retrigger compilation because some parts of the generated binary might be different.

Reply:
That's a flag though and not an environment variable. A build system already needs to handle flags changing. Depinfo tells the build system "hey, I read this environment variable, check it again next time" but in reality it didn't read any environment variable, it got it from the flag.

Reply:
If it is not tracked in the dep info, any change to a `--env` argument would need to trigger recompilation of all crates even if they don't use those `env!()` at all.

Reply:
I see, for build systems that pass env vars explicitly like that it would be nice if they could detect whether it was actually used. But I don't think the normal env depinfo is the right place for that. The case I was thinking about was for example having `A=x` in the environment and using `--env A=x`. Then, changing `A=y` should not cause any rebuilds as it wasn't used, only the `--env A=x` value was.

Reply:
> The case I was thinking about was for example having `A=x` in the environment and using `--env A=x`. Then, changing `A=y` should not cause any rebuilds as it wasn't used, only the `--env A=x` value was.

And in this case it won't trigger a rebuild because the value is same. The `--env` option uses the same mechanism as the `env!` macro: variables are tracked if used and will trigger a rebuild if their value changed. So unless I missed something, I think it's doing exactly what we want?

Reply:
So we're relying on build systems correctly doing the "dep info env -> when I passed --env" translation themselves, since they know they passed the flag.
That makes sense, I didn't think of that, I was too focused on the env depinfo meaning actual environment variable, but I guess the build system is supposed to know what it's doing.
In that case, sounds good. Maybe it should be documented explicitly in the documentation for `--env` that it ends up in depinfo as env vars.